### PR TITLE
Maroon-552 UnitTests for missing UnityEvent targets

### LINKED ISSUE #552: UnitTests to check for missing UnityEvent methods
***Describe the issue***
Currently we have unit tests for missing references via this package introduced here #472 
However, this package cannot find UnityEvent listeners with missing methods. 

***Example how such errors might occur:***
E.g. let's say you connected a button on click event to the method OnButtonClicked() in your script, then you rename that method to OnStartButtonClicked(), now the button cannot find that method anymore. Another example could be that you deleted the object where the script with the method was attached to.

This might look like this (for a Dropdown):
grafik
or like this:
grafik

***Proposed solution***
Write a unit test that checks each experiment scene if any objects have missing methods for UnityEvents (as described here [https://stackoverflow.com/questions/42784338/unity-missing-warning-when-button-has-missing-onclick/42788400?noredirect=1](https://stackoverflow.com/questions/42784338/unity-missing-warning-when-button-has-missing-onclick/42788400?noredirect=1))

### PR DESCRIPTION
**Pull request by FlorianGlawogger**
Fix #552

### CODE REVIEW

File: unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs [hunk 1]
```diff
@@ -371,5 +375,76 @@ public void SceneHasGlobalEntities()
...
+            // Iterate over all MonoBehaviours
+            MonoBehaviour[] monoBehavioursInScene = Resources.FindObjectsOfTypeAll<MonoBehaviour>();
+            foreach(MonoBehaviour monoBehaviour in monoBehavioursInScene)
+            {
+                // Check all fields whether they are a UnityEvent
+                Type monoBehaviourType = monoBehaviour.GetType();
+                FieldInfo[] fields = monoBehaviourType.GetFields(BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public);
+                foreach (FieldInfo field in fields)
+                {
+                    if (typeof(UnityEvent).IsAssignableFrom(field.FieldType))
```
Comment (michaelholly) [inline unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs]: Due to the complex code nesting, I would recommend inverting the if statement.
e.g. if (!typeof(UnityEvent).IsAssignableFrom(field.FieldType)) continue;

File: unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs [hunk 2]
```diff
@@ -371,5 +375,76 @@ public void SceneHasGlobalEntities()
...
+                            Type eventTargetObjectType = Type.GetType(eventTargetObjectFullName);
+                            if (eventTargetObjectType == null)
+                            {
+                                errors.Add($"The UnityEvent of {monoBehaviourType.Name} \"{monoBehaviour.name}\" called \"{field.Name}\" " +
+                                    $"has an event target object whose type is null (index {persistentEventCountIndex}).");
+                                continue;
+                            }
+
+                            // Assert the event target method exists
+                            if (eventTargetObjectType != null)
```
Comment (michaelholly) [inline unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs]: can be inverted.

File: unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs [hunk 3]
```diff
@@ -371,5 +375,76 @@ public void SceneHasGlobalEntities()
...
+                                try
+                                {
+                                    MethodInfo methodInfo = eventTargetObjectType.GetMethod
+                                        (eventTargetMethodName, 
+                                        BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
+                                    if (methodInfo == null)
+                                    {
+                                        errors.Add($"The UnityEvent of  {monoBehaviourType.Name}  \" {monoBehaviour.name} \" called \"{field.Name}\" " +
+                                            $"has an event target method \"{eventTargetMethodName}\" that could not be found (index {persistentEventCountIndex}).");
+                                        continue;
```
Comment (michaelholly) [inline unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs]: I think this line is redundant and can be removed.

File: unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs [hunk 4]
```diff
@@ -371,5 +375,76 @@ public void SceneHasGlobalEntities()
...
+                            // Assert the event target method exists
+                            if (eventTargetObjectType != null)
+                            {
+                                string eventTargetMethodName = unityEvent.GetPersistentMethodName(persistentEventCountIndex);
+                                try
+                                {
+                                    MethodInfo methodInfo = eventTargetObjectType.GetMethod
+                                        (eventTargetMethodName, 
+                                        BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
+                                    if (methodInfo == null)
```
Comment (michaelholly) [inline unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs]: can be inverted too.

File: unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs [hunk 5]
```diff
@@ -371,5 +375,76 @@ public void SceneHasGlobalEntities()
...
+                            UnityEngine.Object eventTargetObject = unityEvent.GetPersistentTarget(persistentEventCountIndex);
+                            if (eventTargetObject == null)
+                            {
+                                errors.Add($"The UnityEvent of {monoBehaviourType.Name} \"{monoBehaviour.name}\" called \"{field.Name}\" " +
+                                    $"has an event target object that is null (index {persistentEventCountIndex}).");
+                                continue;
+                            }
+
+                            // Assert event target object Type is not null
+                            string eventTargetObjectFullName = eventTargetObject.GetType().AssemblyQualifiedName;
```
Comment (michaelholly) [inline unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs]: AssemblyQualifiedName can be null if the current instance represents a generic type parameter.

**Review by michaelholly — CHANGES_REQUESTED** [review]

File: unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs (hunk 5 above)
Comment (FlorianGlawogger) [reply unity/Assets/Tests/EditModeTests/ContentValidation/PcSceneValidationTests.cs]: Not sure I understand this fully;
According to the [docs](https://learn.microsoft.com/en-us/dotnet/api/system.type.assemblyqualifiedname?view=net-8.0) the AssemblyQualifiedName can return null if the type represents a generic type parameter. (like you said)
However, the eventTargetObject should never be a generic Type T? I.e. if I understand correctly, if you have e.g. MyGenericClass<T> myObject and then do typeof(T).AssemblyQualifiedName that might be null? but myObject.GetType().AssemblyQualifiedName should not be null?
But I am not 100% sure I understand that fully, so I'll add a null check, hope that's ok

**Comment by FlorianGlawogger** [comment]
Tests will now fail because now also testing VRScenes, and they fail because of https://github.com/GameLabGraz/Utilities/issues/93 ; so essentially blocked until PR https://github.com/GameLabGraz/Utilities/pull/94 is merged

**Review by michaelholly — APPROVED** [review]

**Pull request merged**